### ISSUE TITLE
Some questions in model.py

### ISSUE THREAD

**Issue by humoxian** [zjunlp/DART#11]
Hello, I am very fond of your paper and work,but I have some problems when I try to understand your code.
In model.py:
`def get_loss(self, batch, full_vocab=True, logits_key='pet_logits'):`
        `# Compute Cross-Entropy loss for prompt verbalizers`
        `assert logits_key in batch, 'logits should be pre-computed and stored in batch dict'`
        `masked_logits = batch[logits_key][batch['pet_flags'] == -1]`
        `labels = batch['pet_labels']`
       ` if not full_vocab:`
            `masked_logits = masked_logits[:, self.label_ids]`
            `labels = batch['label_ids']`
        `return self.loss_fn(masked_logits, labels)`
    

The size of masked_logits is [batch_size,hidden_states] (I used batch_size=4 and bertmodel so the size is [4,768]), and the size of labels is [batch_size,] . However,when computing crossentropy loss,it always raises error that Index out of range. Is there lacking some layers that can transform the masked_logits's size ([4,768]) into [4,num_classes]? 
Meanwhile, I also can't understand the meaning of "full_vocab":when to set it false and when to set it true?
Looking forward to your reply!

**Comment by Riroaki** [comment]
Hi humoxian, refering to the following code, the argument `full_vocab` means whether compute cross entropy loss on all tokens in vocabulary or only on target tokens (i.e., the verbalizer tokens). Empirically we use full vocabulary in loss computation by default for better performance.

https://github.com/zjunlp/DART/blob/d418ded8bd548ef25f2d030990e707a497e93483/src/model.py#L33-L42

FYI, `pet_flags` refers to different types of tokens' positions(not well documented but you can find it [here](https://github.com/zjunlp/DART/blob/d418ded8bd548ef25f2d030990e707a497e93483/src/data.py#LL219C18-L219C70), as `0 = sentence; 1 = special; 2 = prompt; -1 = <mask>`).

Here we get the logits of <mask> tokens, whose size by design should be `[batch, vocab]` or `[batch, class]` (if not `full_vocab`) for our `AutoModelForMaskedLM` model in `run.py`. 

**If the size didn't match here, it is possible that you modified the model class to a base `bertmodel` whose output logit shape could be `[batch, hidden]` like `[4, 768]` as your mentioned.**

For the error `Index out of range`, it is typical when the target label index exceeds the logits' second dimension (i.e., the class dimension). As your logit shape is `[4, 768]` it is possible as the label index might be large (for bert the vocabulary size is ~30,000 tokens).

Hope that helps!

**Comment by humoxian** [comment]
> Hi humoxian, refering to the code
> 
> https://github.com/zjunlp/DART/blob/d418ded8bd548ef25f2d030990e707a497e93483/src/model.py#L33
> 
> , the argument `full_vocab` means whether compute cross entropy loss on all tokens in vocabulary or only on target tokens (i.e., the verbalizer tokens).
> FYI, `pet_flags` refers to different types of tokens' positions(not well documented but you can find it [here](https://github.com/zjunlp/DART/blob/d418ded8bd548ef25f2d030990e707a497e93483/src/data.py#LL219C18-L219C70), as `0 = sentence; 1 = special; 2 = prompt; -1 = <mask>`).
> 
> Here we get the logits of tokens, whose size by design should be `[batch, vocab]` or `[batch, class]` (if not `full_vocab`) for our `AutoModelForMaskedLM` model. **If the size didn't match here, it is possible that you modified the model class to a base `bertmodel` whose output logit shape could be `[batch, hidden]` like `[4, 768]` as your mentioned.**
> 
> For the error `Index out of range`, it is typical when the target label index exceeds the logits' second dimension (i.e., the class dimension). As your logit shape is `[4, 768]` it is possible as the label index might be large (for bert the vocabulary size is ~30,000 tokens).
> 
> Hope that helps!
Okay, I understand now. Thanks very much for your reply!
